### PR TITLE
FIX: improves reliability of post-text-selection bar

### PR DESCRIPTION
**Pull request by jjaffeux**
The main change made is to use `pointerdown` and `touchstart` for detecting click outside in `FloatKit`, the problem of using `click` is that it will trigger on `mouseup` which is not working well with `FloatKit` shown using `mousedown` (when we change selection with the `mousedown` for example) as the release will be interpreted as a click outside and close the menu. To solve this issue the previous code in `post-text-selection` was going through various hacks for detecting state of mouse which are not always very reliable.

The second fix is to exit earlier when selection didn't change.

This has been tested on chrome/firefox and safari (mobile) and seems to work reliably.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/post-text-selection.gjs [hunk 1]
```diff
@@ -89,9 +86,20 @@ export default class PostTextSelection extends Component {
...
+    // avoid hard loops in quote selection unconditionally
+    // this can happen if you triple click text in firefox
+    // it's also generally unecessary work to go
+    // through this if the selection hasn't changed
+    if (this.menuInstance?.expanded && this.prevSelection === _selectedText) {
+      this.prevSelection = _selectedText;
+      return;
+    }
+
+    this.prevSelection = _selectedText;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/post-text-selection.gjs:100]: I move this earlier to exit as soon as possible

File: app/assets/javascripts/float-kit/addon/modifiers/close-on-click-outside.js [hunk 2]
```diff
@@ -12,7 +12,12 @@ export default class FloatKitCloseOnClickOutside extends Modifier {
     this.closeFn = closeFn;
     this.trigger = trigger;
     this.element = element;
-    document.addEventListener("click", this.check, {
+
+    document.addEventListener("pointerdown", this.check, {
+      passive: true,
+    });
+
+    document.addEventListener("touchstart", this.check, {
```
Comment (jjaffeux) [inline app/assets/javascripts/float-kit/addon/modifiers/close-on-click-outside.js:20]: this is the core of the fix

File: app/assets/javascripts/discourse/app/components/post-text-selection.gjs [hunk 3]
```diff
@@ -89,9 +86,20 @@ export default class PostTextSelection extends Component {
...
-    let supportsFastEdit = this.canEditPost;
-    const selection = window.getSelection();
+    const _selectedText = selectedText();
+
+    // avoid hard loops in quote selection unconditionally
+    // this can happen if you triple click text in firefox
+    // it's also generally unecessary work to go
+    // through this if the selection hasn't changed
+    if (this.menuInstance?.expanded && this.prevSelection === _selectedText) {
+      this.prevSelection = _selectedText;
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post-text-selection.gjs:96]: We just checked whether `this.prevSelection === _selectedText`... so is there any need for the assignment here? Seems like a no-op to me?

Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/components/post-text-selection.gjs:96]: yes right I should remove it :+1:

File: app/assets/javascripts/float-kit/addon/modifiers/close-on-click-outside.js (hunk 2 above)
Comment (davidtaylorhq) [reply app/assets/javascripts/float-kit/addon/modifiers/close-on-click-outside.js:20]: Is there a need for `touchstart` here? As I understand it`, pointer events will fire for both mouse and touch devices?

Comment (jjaffeux) [reply app/assets/javascripts/float-kit/addon/modifiers/close-on-click-outside.js:20]: wow 🤯  I was pretty sure it was not the case, TIL, nice.

**Review by davidtaylorhq — APPROVED** [review]

**Pull request merged**